File: devices/kotlin/device/src/main/kotlin/io/realtimeagent/device/StreamChunk.kt
package io.realtimeagent.device

import java.nio.ByteBuffer
import java.nio.ByteOrder

data class StreamChunk(
    val userId: String,
    val sessionId: String,
    val streamId: String,
    val streamType: String,
    val seq: Int,
    val payload: ByteArray,
    val codec: String = "pcm16le",
    val sampleRate: Int = 16000,
    val channels: Int = 1,
    val durationMs: Int = 20,
    val timestampMs: Long = nowMs(),
    val version: String = REALTIME_AGENT_PROTOCOL_VERSION,
    val final: Boolean = false,
    val metadata: Map<String, Any?> = emptyMap(),
) {
    override fun equals(other: Any?): Boolean {
        return other is StreamChunk &&
            userId == other.userId &&
            sessionId == other.sessionId &&
            streamId == other.streamId &&
            streamType == other.streamType &&
            seq == other.seq &&
            payload.contentEquals(other.payload)
    }

    override fun hashCode(): Int = streamId.hashCode()
}

object StreamChunkCodec {
    fun encodeHeader(headerJson: String, payload: ByteArray): ByteArray {
        val headerBytes = headerJson.toByteArray(Charsets.UTF_8)
        return ByteBuffer.allocate(4 + headerBytes.size + payload.size)
            .order(ByteOrder.BIG_ENDIAN)
            .putInt(headerBytes.size)
            .put(headerBytes)
            .put(payload)
            .array()
    }

    fun decodeHeader(raw: ByteArray): Pair<String, ByteArray> {
        require(raw.size >= 4) { "StreamChunk message too short" }
        val headerLength = ByteBuffer.wrap(raw, 0, 4).order(ByteOrder.BIG_ENDIAN).int
        val headerEnd = 4 + headerLength
        require(headerLength > 0 && headerEnd <= raw.size) { "StreamChunk header length is invalid" }
        val headerJson = raw.copyOfRange(4, headerEnd).toString(Charsets.UTF_8)
        val payload = raw.copyOfRange(headerEnd, raw.size)
        val expected = Regex("\"payload_size\"\\s*:\\s*(\\d+)").find(headerJson)?.groupValues?.get(1)?.toInt()
        require(expected == payload.size) { "StreamChunk payload_size mismatch" }
        return headerJson to payload
    }
}

File: devices/javascript/src/stream-chunk.js
import {PROTOCOL_VERSION, nowMs} from "./event.js";

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

export class StreamChunk {
  /**
   * 构造媒体 stream chunk。
   *
   * 主要逻辑：统一 `sensor.mic`、`sensor.rgb` 和 `actuator.speaker` 的二进制信封。
   * 参数：协议头字段和 payload 字节。
   * 返回值：StreamChunk 实例。
   * 异常情况：缺少必要字段时抛出错误。
   */
  constructor({
    userId,
    sessionId,
    streamId,
    streamType,
    seq,
    payload,
    codec,
    sampleRate,
    channels,
    durationMs,
    timestampMs = nowMs(),
    version = PROTOCOL_VERSION,
    final = false,
    metadata = {},
  }) {
    if (!userId || !sessionId || !streamId || !streamType) {
      throw new Error("missing stream chunk identity fields");
    }
    this.userId = userId;
    this.sessionId = sessionId;
    this.streamId = streamId;
    this.streamType = streamType;
    this.seq = seq;
    this.payload = toUint8Array(payload);
    this.codec = codec;
    this.sampleRate = sampleRate;
    this.channels = channels;
    this.durationMs = durationMs;
    this.timestampMs = timestampMs;
    this.version = version;
    this.final = final;
    this.metadata = metadata;
  }
}

export function toUint8Array(value) {
  if (value instanceof Uint8Array) return value;
  if (value instanceof ArrayBuffer) return new Uint8Array(value);
  if (ArrayBuffer.isView(value)) {
    return new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
  }
  return new Uint8Array(value ?? 0);
}

export function encodeStreamChunk(chunk) {
  const normalized = chunk instanceof StreamChunk ? chunk : new StreamChunk(chunk);
  const header = {
    version: normalized.version,
    user_id: normalized.userId,
    session_id: normalized.sessionId,
    stream_id: normalized.streamId,
    stream_type: normalized.streamType,
    seq: normalized.seq,
    timestamp_ms: normalized.timestampMs,
    codec: normalized.codec,
    sample_rate: normalized.sampleRate,
    channels: normalized.channels,
    duration_ms: normalized.durationMs,
    payload_size: normalized.payload.byteLength,
    final: normalized.final,
    metadata: normalized.metadata,
  };
  const headerBytes = textEncoder.encode(JSON.stringify(header));
  const output = new Uint8Array(4 + headerBytes.byteLength + normalized.payload.byteLength);
  const view = new DataView(output.buffer);
  view.setUint32(0, headerBytes.byteLength, false);
  output.set(headerBytes, 4);
  output.set(normalized.payload, 4 + headerBytes.byteLength);
  return output;
}

export function decodeStreamChunk(data) {
  const bytes = toUint8Array(data);
  if (bytes.byteLength < 4) {
    throw new Error("stream chunk 格式错误：message too short");
  }
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const headerLength = view.getUint32(0, false);
  const headerEnd = 4 + headerLength;
  if (headerLength <= 0 || headerEnd > bytes.byteLength) {
    throw new Error("stream chunk 格式错误：invalid header length");
  }
  const header = JSON.parse(textDecoder.decode(bytes.slice(4, headerEnd)));
  const payload = bytes.slice(headerEnd);
  if (header.payload_size !== payload.byteLength) {
    throw new Error("stream chunk 格式错误：payload_size mismatch");
  }
  return new StreamChunk({
    userId: requiredString(header, "user_id"),
    sessionId: requiredString(header, "session_id"),
    streamId: requiredString(header, "stream_id"),
    streamType: requiredString(header, "stream_type"),
    seq: requiredNumber(header, "seq"),
    payload,
    codec: requiredString(header, "codec"),
    sampleRate: requiredNumber(header, "sample_rate"),
    channels: requiredNumber(header, "channels"),
    durationMs: requiredNumber(header, "duration_ms"),
    timestampMs: requiredNumber(header, "timestamp_ms"),
    version: header.version ?? PROTOCOL_VERSION,
    final: Boolean(header.final),
    metadata: header.metadata ?? {},
  });
}

function requiredString(header, key) {
  const value = header[key];
  if (typeof value !== "string") throw new Error(`stream chunk 格式错误：missing ${key}`);
  return value;
}

function requiredNumber(header, key) {
  const value = header[key];
  if (typeof value !== "number") throw new Error(`stream chunk 格式错误：missing ${key}`);
  return value;
}
